### PR TITLE
[libc++] Deprecate the _LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS macro

### LINKED ISSUE #63900: _LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS causes bad diagnostics
I'm upgrading Meta's Android codebase to C++20, and I'm building with `-D_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS` globally to ease the initial transition. That somehow causes incomplete backtraces in some diagnostics. See https://godbolt.org/z/TMKWrzd59 as an example, where the error message has no reference to the actual problematic `emplace_back` call, whereas either building with `-std=c++17` (https://godbolt.org/z/1oq4q4Pb8) or removing `-D_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS` (https://godbolt.org/z/6v3f7aKor) gives the expected full error.

Can we improve this? I expect it's not uncommon to use `-D_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS` and other definitions during the C++20 upgrade process, and there are certain errors that are exposed only when building with C++20 (e.g. https://godbolt.org/z/cEMvdrca8), so this issue makes it much harder to figure out the cause of those errors.

### PR DESCRIPTION
**Pull request by ldionne**
As described in #69994, using the escape hatch makes us non-conforming in C++20 due to incorrect constexpr-ness. It also leads to bad diagnostics as reported by #63900. We discussed the issue in the libc++ monthly meeting and we agreed that we should deprecate the macro in LLVM 18, and then remove it in LLVM 19 since it causes too many problems.

This patch does the first part of this -- it deprecates the macro.

Fixes #69994
Fixes #63900
Partially addresses #75975

### CODE REVIEW

File: libcxx/docs/ReleaseNotes/18.rst [hunk 1]
```diff
@@ -117,6 +117,12 @@ Deprecations and Removals
   and ``<experimental/vector>`` have been removed in LLVM 18, as all their contents will have been
   implemented in namespace ``std`` for at least two releases.
 
+- The macro ``_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS`` has been deprecated and will be removed
+  in LLVM 19. This macro used to re-enable redundant members of ``std::allocator<T>`` like ``pointer``,
+  ``reference``, ``rebind``, ``address``, ``max_size``, ``construct``, ``destroy``, and the two-argument
+  overload of ``allocate``. However, this led to the library being non-conforming due to incorrect
+  constexpr-ness. As a result the macro is being deprecated and will be removed in LLVM 19.
```
Comment (mordante) [inline libcxx/docs/ReleaseNotes/18.rst:124]: The last sentence duplicates the first in this hunk.
```suggestion
  constexpr-ness.
```

**Review by mordante — APPROVED** [review]
LGTM modulo one nit and when the CI is green.

**Comment by jyknight** [comment]
My suggestion on #69994 had been to stop implying `_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS` from `_LIBCPP_ENABLE_CXX20_REMOVED_FEATURES` in LLVM 18 at the same time as deprecating it. Did you intend to _not_ do that, or was it just missed?

**Comment by philnik777** [comment]
> My suggestion on #69994 had been to stop implying `_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS` from `_LIBCPP_ENABLE_CXX20_REMOVED_FEATURES` in LLVM 18 at the same time as deprecating it. Did you intend to _not_ do that, or was it just missed?

We've also deprecated (or plan to?) `_LIBCPP_ENABLE_CXX20_REMOVED_FEATURES`, so there isn't really a point in changing anything there now.

**Comment by jyknight** [comment]
Ah, I see that now. SGTM, thanks for the clarification!

**Comment by mordante** [comment]
> > My suggestion on #69994 had been to stop implying `_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS` from `_LIBCPP_ENABLE_CXX20_REMOVED_FEATURES` in LLVM 18 at the same time as deprecating it. Did you intend to _not_ do that, or was it just missed?
> 
> We've also deprecated (or plan to?) `_LIBCPP_ENABLE_CXX20_REMOVED_FEATURES`, so there isn't really a point in changing anything there now.

I've created https://github.com/llvm/llvm-project/pull/77879 for that yesterday.

**Pull request merged**